src/engine/QueryExecutionTree.cpp

```diff
@@ -171,34 +170,17 @@ QueryExecutionTree::createSortedTreeAnyPermutation(
 std::shared_ptr<QueryExecutionTree> QueryExecutionTree::createSortedTree(
     std::shared_ptr<QueryExecutionTree> qet,
     const vector<ColumnIndex>& sortColumns) {
-  auto inputSortedOn = qet->resultSortedOn();
-  bool inputSorted = sortColumns.size() <= inputSortedOn.size();
-  for (size_t i = 0; inputSorted && i < sortColumns.size(); ++i) {
-    inputSorted = sortColumns[i] == inputSortedOn[i];
-  }
-  if (sortColumns.empty() || inputSorted) {
-    return qet;
-  }
-
-  // Unwrap sort to avoid stacking sorts on top of each other.
-  if (auto sort = std::dynamic_pointer_cast<Sort>(qet->getRootOperation())) {
-    AD_LOG_DEBUG << "Tried to re-sort a subtree that will already be sorted "
-                    "with `Sort` with a different sort order. This is "
-                    "indicates a flaw during query planning."
-                 << std::endl;
-    qet = sort->getSubtree();
-  }
-
-  // Push down sort into Union.
-  QueryExecutionContext* qec = qet->getRootOperation()->getExecutionContext();
-  if (auto unionOperation =
-          std::dynamic_pointer_cast<Union>(qet->getRootOperation())) {
-    return std::make_shared<QueryExecutionTree>(
-        qec, unionOperation->createSortedVariant(sortColumns));
-  }
-
-  auto sort = std::make_shared<Sort>(qec, std::move(qet), sortColumns);
-  return std::make_shared<QueryExecutionTree>(qec, std::move(sort));
+  const auto& rootOperation = qet->getRootOperation();
+  auto* qec = rootOperation->getExecutionContext();
+  PreconditionAction state = rootOperation->createSortedClone(sortColumns);
+
+  return std::move(state)
+      .handle([qec, &qet, &sortColumns]() mutable {
+        auto sort = std::make_shared<Sort>(qec, std::move(qet), sortColumns);
+        return std::make_shared<QueryExecutionTree>(qec, std::move(sort));
+      })
+      .getTree()
+      .value_or(qet);
```

Comment:
Maybe (depending on how common the pattern is), 
you can merge those three things to a function. `satisfy(inputTree, functionFromTreeToTree)`.

Reply:
It is not common enough to make it worth it.